api/interceptor/observer.go

```diff
@@ -0,0 +1,269 @@
+package interceptor
+
+import (
+	"context"
+	"net/http"
+	"strings"
+	"time"
+
+	"connectrpc.com/connect"
+	"github.com/grpc-ecosystem/grpc-gateway/v2/runtime"
+)
+
+var _ connect.Interceptor = &Observer{}
+
+// Observer wraps all requests/responses and emits events that inform about
+// request duration, size and status code.
+type Observer struct {
+	onRequestEnded func(context.Context, *RequestMetadata)
+}
+
+// NewObserver creates a new Observer.
+func NewObserver(onRequestEnded func(context.Context, *RequestMetadata)) *Observer {
+	return &Observer{
+		onRequestEnded: onRequestEnded,
+	}
+}
+
+// WrapHandler mounts an HTTP middleware to gather metrics at the
+// HTTP level.
+func (in *Observer) WrapHandler(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// 1. Inject request metadata into request context
+		var procedure string
+		switch {
+		case strings.HasPrefix(r.RequestURI, "/grpc.reflection.v1.ServerReflection"),
+			strings.HasPrefix(r.RequestURI, "/grpc.reflection.v1alpha.ServerReflection"):
+			procedure = r.RequestURI
+		default:
+			procedure = "none"
+		}
+		rm := &RequestMetadata{
+			startAt:     time.Now(),
+			peerAddress: r.RemoteAddr,
+			protocol:    "http",
```

Comment:
Should we use `r.Proto` ?

Reply:
I'm not sure, my main intention was to differentiate between http, connectrpc and grpc. Do you think we should be more precise about the different http protocols?

Reply:
It was just a thought. I am fine with this too.